File: nodejsts/src/controllers/ClientIntakeController.ts
import { Request, Response } from 'express';
import { ClientIntakeService } from '../services/ClientIntakeService';
import { asyncHandler } from '../middleware/errorHandler';
import { ApiResponse } from '../types';

export class ClientIntakeController {
  private clientIntakeService: ClientIntakeService;

  constructor() {
    this.clientIntakeService = new ClientIntakeService();
  }

  createClientIntake = asyncHandler(async (req: Request, res: Response): Promise<void> => {
    const clientIntake = await this.clientIntakeService.createClientIntake(req.body);
    const response: ApiResponse = {
      success: true,
      data: clientIntake,
      message: 'Client intake created successfully'
    };
    res.status(201).json(response);
  });

  getAllClientIntakes = asyncHandler(async (req: Request, res: Response): Promise<void> => {
    const clientIntakes = await this.clientIntakeService.getAllClientIntakes();
    const response: ApiResponse = {
      success: true,
      data: clientIntakes,
      message: 'Client intakes retrieved successfully'
    };
    res.status(200).json(response);
  });

  getClientIntakeById = asyncHandler(async (req: Request, res: Response): Promise<void> => {
    const clientIntakeId = req.params.id;
    if (!clientIntakeId) {
      res.status(400).json({ success: false, message: 'Client intake ID is required' });
      return;
    }

    const clientIntake = await this.clientIntakeService.getClientIntakeById(clientIntakeId);
    const response: ApiResponse = {
      success: true,
      data: clientIntake,
      message: 'Client intake retrieved successfully'
    };
    res.status(200).json(response);
  });

  updateClientIntake = asyncHandler(async (req: Request, res: Response): Promise<void> => {
    const clientIntakeId = req.params.id;
    if (!clientIntakeId) {
      res.status(400).json({ success: false, message: 'Client intake ID is required' });
      return;
    }

    const clientIntake = await this.clientIntakeService.updateClientIntake(clientIntakeId, req.body);
    const response: ApiResponse = {
      success: true,
      data: clientIntake,
      message: 'Client intake updated successfully'
    };
    res.status(200).json(response);
  });

  deleteClientIntake = asyncHandler(async (req: Request, res: Response): Promise<void> => {
    const clientIntakeId = req.params.id;
    if (!clientIntakeId) {
      res.status(400).json({ success: false, message: 'Client intake ID is required' });
      return;
    }

    const result = await this.clientIntakeService.deleteClientIntake(clientIntakeId);
    const response: ApiResponse = {
      success: true,
      data: result,
      message: 'Client intake deleted successfully'
    };
    res.status(200).json(response);
  });
}

File: JavaSpringBoot/src/main/java/com/whoestate/controller/ClientIntakeController.java
package com.whoestate.controller;

import java.util.List;

import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.http.HttpStatus;
import org.springframework.http.ResponseEntity;
import org.springframework.security.access.prepost.PreAuthorize;
import org.springframework.web.bind.annotation.CrossOrigin;
import org.springframework.web.bind.annotation.DeleteMapping;
import org.springframework.web.bind.annotation.GetMapping;
import org.springframework.web.bind.annotation.PathVariable;
import org.springframework.web.bind.annotation.PostMapping;
import org.springframework.web.bind.annotation.RequestBody;
import org.springframework.web.bind.annotation.RequestMapping;
import org.springframework.web.bind.annotation.RestController;

import com.whoestate.dto.CreateClientIntakeDto;
import com.whoestate.entity.ClientIntake;
import com.whoestate.service.ClientIntakeService;

import jakarta.validation.Valid;

@RestController
@RequestMapping("/api/client-intakes")
@CrossOrigin(origins = "*")
public class ClientIntakeController {

    @Autowired
    private ClientIntakeService clientIntakeService;

    @GetMapping
    @PreAuthorize("hasRole('Admin')")
    public ResponseEntity<List<ClientIntake>> getAllClientIntakes() {
        List<ClientIntake> clientIntakes = clientIntakeService.findAll();
        return ResponseEntity.ok(clientIntakes);
    }

    @GetMapping("/{id}")
    public ResponseEntity<ClientIntake> getClientIntake(@PathVariable String id) {
        ClientIntake clientIntake = clientIntakeService.findById(id);
        if (clientIntake != null) {
            return ResponseEntity.ok(clientIntake);
        } else {
            return ResponseEntity.notFound().build();
        }
    }

    @PostMapping
    public ResponseEntity<ClientIntake> createClientIntake(@Valid @RequestBody CreateClientIntakeDto createClientIntakeDto) {
        ClientIntake clientIntake = clientIntakeService.create(createClientIntakeDto);
        return ResponseEntity.status(HttpStatus.CREATED).body(clientIntake);
    }

    @DeleteMapping("/{id}")
    @PreAuthorize("hasRole('Admin')")
    public ResponseEntity<?> deleteClientIntake(@PathVariable String id) {
        boolean success = clientIntakeService.delete(id);
        if (success) {
            return ResponseEntity.ok(new Object() {
                public String message = "Client intake deleted successfully";
            });
        } else {
            return ResponseEntity.notFound().build();
        }
    }
}